contracts/multisig-prover/src/execute.rs

```diff
@@ -142,64 +142,115 @@ fn get_workers_info(
     })
 }
 
-pub fn update_worker_set(deps: DepsMut, env: Env) -> Result<Response, ContractError> {
-    let config = CONFIG.load(deps.storage)?;
-
-    let workers_info = get_workers_info(&deps, &env, &config)?;
+fn get_next_worker_set(
+    deps: &DepsMut,
+    env: &Env,
+    config: &Config,
+) -> Result<Option<WorkerSet>, ContractError> {
+    let workers_info = get_workers_info(deps, env, config)?;
 
+    let cur_worker_set = CURRENT_WORKER_SET.may_load(deps.storage)?;
     let new_worker_set = WorkerSet::new(
         workers_info.pubkeys_by_participant,
         workers_info.snapshot.quorum.into(),
         env.block.height,
     )?;
 
+    match cur_worker_set {
+        Some(cur_worker_set) => {
+            if should_update_worker_set(
+                &new_worker_set,
+                &cur_worker_set,
+                config.worker_set_diff_threshold as usize,
+            ) {
+                Ok(Some(new_worker_set))
+            } else {
+                Ok(None)
+            }
+        }
+        None => Ok(Some(new_worker_set)),
+    }
+}
+
+fn save_next_worker_set(
+    storage: &mut dyn Storage,
+    workers_info: WorkersInfo,
+    new_worker_set: WorkerSet,
+) -> Result<(), ContractError> {
+    if different_set_in_progress(storage, &new_worker_set) {
+        return Err(ContractError::WorkerSetConfirmationInProgress);
+    }
+
+    Ok(NEXT_WORKER_SET.save(storage, &(new_worker_set, workers_info.snapshot))?)
+}
+
+fn initialize_worker_set(
+    storage: &mut dyn Storage,
+    new_worker_set: WorkerSet,
+) -> Result<(), ContractError> {
+    let key_id = new_worker_set.id(); // this is really just the worker_set_id
+
+    CURRENT_WORKER_SET.save(storage, &new_worker_set)?;
+    KEY_ID.save(storage, &key_id)?;
+
+    Ok(())
+}
+
+fn make_keygen_msg(
+    key_id: String,
+    snapshot: axelar_wasm_std::Snapshot,
+    worker_set: WorkerSet,
+) -> multisig::msg::ExecuteMsg {
+    multisig::msg::ExecuteMsg::KeyGen {
+        key_id,
+        snapshot,
+        pub_keys_by_address: worker_set
+            .signers
+            .into_iter()
+            .map(|signer| {
+                (
+                    signer.address.to_string(),
+                    (KeyType::Ecdsa, signer.pub_key.as_ref().into()),
+                )
+            })
+            .collect(),
+    }
+}
+
+pub fn update_worker_set(deps: DepsMut, env: Env) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    let workers_info = get_workers_info(&deps, &env, &config)?;
     let cur_worker_set = CURRENT_WORKER_SET.may_load(deps.storage)?;
+    let new_worker_set = get_next_worker_set(&deps, &env, &config)?;
 
-    let key_id = new_worker_set.id();
+    if new_worker_set.is_none() {
+        return Err(ContractError::WorkerSetUnchanged);
+    }
+
+    //.unwrap() is safe here because we know new_worker_set is not none from the if statement above.
+    let new_worker_set_unwrapped = new_worker_set.unwrap();
```

Comment:
.unwrap() is safe here because the conditional above ensures it isn't a None.

? does not work here for Option's because the function update_worker_set returns a Result.